File: start-server.js
#!/usr/bin/env node

import { spawn } from 'child_process';
import { existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

class ServerLauncher {
  constructor() {
    this.preferredRuntime = process.env.MCP_RUNTIME || 'auto';
  }

  async checkRuntime(command) {
    return new Promise((resolve) => {
      const child = spawn(command, ['--version'], { stdio: 'ignore' });
      child.on('close', (code) => resolve(code === 0));
      child.on('error', () => resolve(false));
    });
  }

  async detectAvailableRuntimes() {
    const runtimes = {
      node: await this.checkRuntime('node'),
      python: await this.checkRuntime('python') || await this.checkRuntime('python3')
    };
    
    console.error('Available runtimes:', Object.entries(runtimes)
      .filter(([_, available]) => available)
      .map(([runtime]) => runtime)
      .join(', ') || 'none');
    
    return runtimes;
  }

  async startServer() {
    const runtimes = await this.detectAvailableRuntimes();
    let serverCommand, serverArgs, serverFile;

    // Determine which server to run
    if (this.preferredRuntime === 'python' && runtimes.python) {
      const pythonCmd = await this.checkRuntime('python3') ? 'python3' : 'python';
      serverCommand = pythonCmd;
      serverArgs = ['main.py'];
      serverFile = 'Python';
    } else if (this.preferredRuntime === 'node' && runtimes.node) {
      serverCommand = 'node';
      serverArgs = ['mcp-server.js'];
      serverFile = 'Node.js';
    } else if (this.preferredRuntime === 'auto') {
      // Auto-detect: prefer Node.js, fallback to Python
      if (runtimes.node) {
        serverCommand = 'node';
        serverArgs = ['mcp-server.js'];
        serverFile = 'Node.js';
      } else if (runtimes.python) {
        const pythonCmd = await this.checkRuntime('python3') ? 'python3' : 'python';
        serverCommand = pythonCmd;
        serverArgs = ['main.py'];
        serverFile = 'Python';
      }
    }

    if (!serverCommand) {
      console.error('Error: No suitable runtime found or specified runtime not available');
      console.error('Available options:');
      if (runtimes.node) console.error('  - Set MCP_RUNTIME=node to use Node.js');
      if (runtimes.python) console.error('  - Set MCP_RUNTIME=python to use Python');
      process.exit(1);
    }

    console.error(`Starting DealMate MCP Server using ${serverFile}...`);
    
    // Start the server
    const server = spawn(serverCommand, serverArgs, {
      cwd: __dirname,
      stdio: 'inherit',
      env: { ...process.env }
    });

    server.on('error', (error) => {
      console.error(`Failed to start server: ${error.message}`);
      process.exit(1);
    });

    server.on('close', (code) => {
      console.error(`Server exited with code ${code}`);
      process.exit(code);
    });

    // Handle graceful shutdown
    process.on('SIGINT', () => {
      console.error('Shutting down server...');
      server.kill('SIGINT');
    });

    process.on('SIGTERM', () => {
      console.error('Shutting down server...');
      server.kill('SIGTERM');
    });
  }
}

const launcher = new ServerLauncher();
launcher.startServer().catch(console.error);

File: start-server.py
#!/usr/bin/env python3

import os
import sys
import subprocess
import shutil
from pathlib import Path

class ServerLauncher:
    def __init__(self):
        self.preferred_runtime = os.getenv('MCP_RUNTIME', 'auto')
        self.script_dir = Path(__file__).parent
    
    def check_runtime(self, command):
        """Check if a runtime is available"""
        try:
            result = subprocess.run([command, '--version'], 
                                  capture_output=True, timeout=5)
            return result.returncode == 0
        except (subprocess.TimeoutExpired, FileNotFoundError):
            return False
    
    def detect_available_runtimes(self):
        """Detect available runtimes"""
        runtimes = {
            'node': self.check_runtime('node'),
            'python': self.check_runtime('python') or self.check_runtime('python3')
        }
        
        available = [runtime for runtime, available in runtimes.items() if available]
        print(f"Available runtimes: {', '.join(available) if available else 'none'}", 
              file=sys.stderr)
        
        return runtimes
    
    def start_server(self):
        """Start the appropriate server"""
        runtimes = self.detect_available_runtimes()
        server_command = None
        server_args = []
        server_file = None
        
        # Determine which server to run
        if self.preferred_runtime == 'python' and runtimes['python']:
            python_cmd = 'python3' if self.check_runtime('python3') else 'python'
            server_command = python_cmd
            server_args = ['main.py']
            server_file = 'Python'
        elif self.preferred_runtime == 'node' and runtimes['node']:
            server_command = 'node'
            server_args = ['mcp-server.js']
            server_file = 'Node.js'
        elif self.preferred_runtime == 'auto':
            # Auto-detect: prefer Node.js, fallback to Python
            if runtimes['node']:
                server_command = 'node'
                server_args = ['mcp-server.js']
                server_file = 'Node.js'
            elif runtimes['python']:
                python_cmd = 'python3' if self.check_runtime('python3') else 'python'
                server_command = python_cmd
                server_args = ['main.py']
                server_file = 'Python'
        
        if not server_command:
            print("Error: No suitable runtime found or specified runtime not available", 
                  file=sys.stderr)
            print("Available options:", file=sys.stderr)
            if runtimes['node']:
                print("  - Set MCP_RUNTIME=node to use Node.js", file=sys.stderr)
            if runtimes['python']:
                print("  - Set MCP_RUNTIME=python to use Python", file=sys.stderr)
            sys.exit(1)
        
        print(f"Starting DealMate MCP Server using {server_file}...", file=sys.stderr)
        
        # Start the server
        try:
            subprocess.run([server_command] + server_args, cwd=self.script_dir)
        except KeyboardInterrupt:
            print("Server shutdown requested", file=sys.stderr)
        except Exception as e:
            print(f"Failed to start server: {e}", file=sys.stderr)
            sys.exit(1)

if __name__ == '__main__':
    launcher = ServerLauncher()
    launcher.start_server()